source/fe/fe_p1nc.cc

```diff
@@ -0,0 +1,418 @@
+                                 const InternalDataBase                                              &fe_internal,
+                                 dealii::internal::FEValues::FiniteElementRelatedData<2,2> &output_data) const
+
+{
+  const UpdateFlags flags(fe_internal.update_each) ;
+
+  const unsigned int n_q_points = mapping_data.quadrature_points.size();
+
+  std::vector<double> values(flags & update_values ? this->dofs_per_cell : 0);
+  std::vector<Tensor<1,2> > grads(flags & update_gradients ? this->dofs_per_cell : 0);
+
+
+  // linear shape with a half value: phi(x,y) = ax + by + c
+  std::vector<double> a(4), b(4), c(4);
+  get_linear_shape (cell, a, b, c);
+
+
+  // compute basis functions
+  if (flags & (update_values | update_gradients))
+    for (unsigned int i=0; i<n_q_points; ++i)
+      {
+        for (unsigned int k=0; k<this->dofs_per_cell; ++k)
+          {
+            if (flags & update_values)
+              {
+                values[k] = a[k]*mapping_data.quadrature_points[i](0) + b[k]*mapping_data.quadrature_points[i](1) + c[k] ;
+                output_data.shape_values[k][i] = values[k];
+              }
+
+            if (flags & update_gradients)
+              {
+                grads[k][0] = a[k] ;
+                grads[k][1] = b[k] ;
+                output_data.shape_gradients[k][i] = grads[k];
+              }
+          }
+      }
+
+  // When this function is called for computation of facial jump residual,
+  // MappingRelatedData does not work properly.
+  // In this case, the quadrature points on the real cell is computed in manual sense, using 'mapping' and 'quadrature'.
+  // This is a temporary solution. It needs to be fixed fundamentally.
+  if (n_q_points==0)
+    {
+      Quadrature<2> cellquadrature = QProjector<2>::project_to_subface(quadrature, face_no, sub_no) ;
+      for (unsigned int i=0; i<cellquadrature.size(); ++i)
+        for (unsigned int k=0; k<this->dofs_per_cell; ++k)
+          {
+            if (flags & update_values)
+              {
+                Point<2> realquadrature ;
+
+                realquadrature = mapping.transform_unit_to_real_cell(cell, cellquadrature.point(i)) ;
+                values[k] = a[k]*realquadrature(0) + b[k]*realquadrature(1) + c[k] ;
+                output_data.shape_values[k][i] = values[k];
+              }
+
+            if (flags & update_gradients)
+              {
+                grads[k][0] = a[k] ;
+                grads[k][1] = b[k] ;
+                output_data.shape_gradients[k][i] = grads[k];
+              }
+          }
+    }
+}
+
+
+
+void FE_P1NC::initialize_constraints ()
+{
+
+  std::vector<Point<1> > constraint_points;
+
+  // Add midpoint
+  constraint_points.push_back (Point<1> (0.5));
+
+  // coefficient relation between children and mother
+  interface_constraints
+  .TableBase<2,double>::reinit (interface_constraints_size());
+
+  interface_constraints(0,0) = 0.5 ;
+  interface_constraints(0,1) = 0.5 ;
```

Comment:
OK, thanks for checking.

I think ultimately it would be nice to add a test to the testsuite that demonstrates convergence of the element, and a second one that demonstrates convergence on an adaptively refined mesh. I bet you already have such a code anyway. But that can wait until all of the other issues are sorted out.